### PR TITLE
feat: Add Firestore module

### PR DESCRIPTION
**Pull request by dejandjenic**
## Related issues

-

### CODE REVIEW

File: Testcontainers.sln [hunk 1]
```diff
@@ -13,138 +13,139 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "src", "src", "{673F23AE-769
 EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "tests", "tests", "{7164F1FB-7F24-444A-ACD2-2C329C2B3CCF}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Testcontainers.Azurite", "src\Testcontainers.Azurite\Testcontainers.Azurite.csproj", "{3F2E254F-C203-43FD-A078-DC3E2CBC0F9F}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Testcontainers.Azurite", "src\Testcontainers.Azurite\Testcontainers.Azurite.csproj", "{3F2E254F-C203-43FD-A078-DC3E2CBC0F9F}"
```
Comment (HofmeisterAn) [inline Testcontainers.sln]: Please remove (all) the unnecessary changes in the Solution (.sln) file.

File: hooks/post-checkout [hunk 2]
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
```
Comment (HofmeisterAn) [inline hooks/post-checkout]: I remember you mentioned you have issues using Git LFS. I am not sure what happened here, but please remove the hooks and binary files.

File: src/Testcontainers.Firestore/FirestoreBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,74 @@
+namespace Testcontainers.Firestore;
+
+/// <inheritdoc cref="ContainerBuilder{TBuilderEntity, TContainerEntity, TConfigurationEntity}" />
+[PublicAPI]
+public sealed class FirestoreBuilder : ContainerBuilder<FirestoreBuilder, FirestoreContainer, FirestoreConfiguration>
+{
+    const string Command = "gcloud beta emulators firestore start --host-port 0.0.0.0:8080";
```
Comment (HofmeisterAn) [inline src/Testcontainers.Firestore/FirestoreBuilder.cs]: Please use the exec form in `Init()`, like:

    .WithEntrypoint("gcloud")
    .WithCommand("beta", "emulators", "firestore", "start", "--host-port", "0.0.0.0:" + FirestorePort)

File: src/Testcontainers.Firestore/FirestoreBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        return new FirestoreBuilder(new FirestoreConfiguration(oldValue, newValue));
+    }
+
+
+    /// <inheritdoc />
+    protected override FirestoreBuilder Init()
+    {
+        return base.Init()
+          .WithImage(Image)
+          .WithPortBinding(8080)
```
Comment (HofmeisterAn) [inline src/Testcontainers.Firestore/FirestoreBuilder.cs]: Please use random assigned ports as mentioned in our [best practices](https://dotnet.testcontainers.org/api/best_practices/).

```suggestion
          .WithPortBinding(8080, true)
```

File: src/Testcontainers.Firestore/FirestoreBuilder.cs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+          .WithCommand("/bin/sh", "-c", Command)          
+          ;
+    }
+
+    /// <inheritdoc />
+    protected override void Validate()
+    {
+        base.Validate();
+    }
+
```
Comment (HofmeisterAn) [inline src/Testcontainers.Firestore/FirestoreBuilder.cs]:
```suggestion
```

File: src/Testcontainers.Firestore/FirestoreConfiguration.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    /// Initializes a new instance of the <see cref="FirestoreConfiguration" /> class.
+    /// </summary>
+    /// <param name="oldValue">The old Docker resource configuration.</param>
+    /// <param name="newValue">The new Docker resource configuration.</param>
+    public FirestoreConfiguration(FirestoreConfiguration oldValue, FirestoreConfiguration newValue)
+        : base(oldValue, newValue)
+    {
+        // // Create an updated immutable copy of the module configuration.
+    }
+
```
Comment (HofmeisterAn) [inline src/Testcontainers.Firestore/FirestoreConfiguration.cs]:
```suggestion
```

File: src/Testcontainers.Firestore/FirestoreContainer.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    /// <param name="configuration">The container configuration.</param>
+    /// <param name="logger">The logger.</param>
+    public FirestoreContainer(FirestoreConfiguration configuration, ILogger logger)
+        : base(configuration, logger)
+    {
+    }
+
+    public void SetEmulatorHost()
+    {
+        Environment.SetEnvironmentVariable("FIRESTORE_EMULATOR_HOST", $"{Hostname}:8080/");
```
Comment (HofmeisterAn) [inline src/Testcontainers.Firestore/FirestoreContainer.cs]: As previously requested (in the other pull request), can we please not rely on the environment variable to create a client connection? Environment variables are shared (min. across the process) and prevents developers from running multiple instances concurrently.

File: tests/Testcontainers.Firestore.Tests/Testcontainers.Firestore.Tests.csproj [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+      <PackageReference Include="xunit.runner.visualstudio" Version="2.4.5"/>
+      <PackageReference Include="xunit" Version="2.4.2"/>
+      <PackageReference Include="Google.Cloud.Firestore" Version="3.1.0" />
+    </ItemGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="$(SolutionDir)src/Testcontainers.Firestore/Testcontainers.Firestore.csproj"/>
+    <ProjectReference Include="$(SolutionDir)tests/Testcontainers.Commons/Testcontainers.Commons.csproj"/>
+  </ItemGroup>
+
```
Comment (HofmeisterAn) [inline tests/Testcontainers.Firestore.Tests/Testcontainers.Firestore.Tests.csproj]:
```suggestion
    <PropertyGroup>
        <TargetFramework>net6.0</TargetFramework>
        <IsPackable>false</IsPackable>
        <IsPublishable>false</IsPublishable>
    </PropertyGroup>
    <ItemGroup>
        <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.4.1"/>
        <PackageReference Include="coverlet.collector" Version="3.2.0"/>
        <PackageReference Include="xunit.runner.visualstudio" Version="2.4.5"/>
        <PackageReference Include="xunit" Version="2.4.2"/>
        <PackageReference Include="Google.Cloud.Firestore" Version="3.1.0"/>
    </ItemGroup>
    <ItemGroup>
        <ProjectReference Include="$(SolutionDir)src/Testcontainers.Firestore/Testcontainers.Firestore.csproj"/>
        <ProjectReference Include="$(SolutionDir)tests/Testcontainers.Commons/Testcontainers.Commons.csproj"/>
    </ItemGroup>
```

File: tests/Testcontainers.Firestore.Tests/FirestoreContainerTests.cs [hunk 9]
```diff
@@ -0,0 +1,59 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using DotNet.Testcontainers.Commons;
+using Google.Api.Gax;
+using Google.Apis.Http;
+using Google.Cloud.Firestore;
```
Comment (HofmeisterAn) [inline tests/Testcontainers.Firestore.Tests/FirestoreContainerTests.cs]: Move usings to global.

**Review by HofmeisterAn — CHANGES_REQUESTED** [review]
Thank you for your pull request. I have some requests. As soon as they are addressed, we can merge the module. Thanks.

File: tests/Testcontainers.Firestore.Tests/FirestoreContainerTests.cs (hunk 9 above)
Comment (dejandjenic) [reply tests/Testcontainers.Firestore.Tests/FirestoreContainerTests.cs]: done

File: tests/Testcontainers.Firestore.Tests/Testcontainers.Firestore.Tests.csproj (hunk 8 above)
Comment (dejandjenic) [reply tests/Testcontainers.Firestore.Tests/Testcontainers.Firestore.Tests.csproj]: done

File: src/Testcontainers.Firestore/FirestoreContainer.cs (hunk 7 above)
Comment (dejandjenic) [reply src/Testcontainers.Firestore/FirestoreContainer.cs]: done

File: src/Testcontainers.Firestore/FirestoreConfiguration.cs (hunk 6 above)
Comment (dejandjenic) [reply src/Testcontainers.Firestore/FirestoreConfiguration.cs]: done

File: src/Testcontainers.Firestore/FirestoreBuilder.cs (hunk 5 above)
Comment (dejandjenic) [reply src/Testcontainers.Firestore/FirestoreBuilder.cs]: done

File: src/Testcontainers.Firestore/FirestoreBuilder.cs (hunk 4 above)
Comment (dejandjenic) [reply src/Testcontainers.Firestore/FirestoreBuilder.cs]: done

File: src/Testcontainers.Firestore/FirestoreBuilder.cs (hunk 3 above)
Comment (dejandjenic) [reply src/Testcontainers.Firestore/FirestoreBuilder.cs]: done

File: hooks/post-checkout (hunk 2 above)
Comment (dejandjenic) [reply hooks/post-checkout]: done

File: Testcontainers.sln (hunk 1 above)
Comment (dejandjenic) [reply Testcontainers.sln]: done

**Comment by dejandjenic** [comment]
i pushed new changes to resolve comments made to PR, i didn't resolve comments themselves unsure what are practices here. if required i can do that.

**Review by HofmeisterAn — COMMENTED** [review]
Thank you for updating the pull request and addressing the requests. I made some minor adjustments, nothing significant. The pull request looks good 👍.

One thing I'd like to discuss is whether it makes sense to rename the module to "GCloud," similar to Java. This change would allow us to provide different container implementations, much like what Java does [here](https://github.com/testcontainers/testcontainers-java/tree/main/modules/gcloud/src/main/java/org/testcontainers/containers). Additionally, should we renaming the container to `FirestoreEmulatorContainer`? This PR is related to https://github.com/testcontainers/testcontainers-dotnet/issues/862 too.

**Comment by dejandjenic** [comment]
I can do a push with renaming tommorow, gcloud naming is similar to java containers but dotnet containers have different setup. E. G. Azure is not single module. What ever you decide. I plan to push pubsub in aonther pr also

**Comment by HofmeisterAn** [comment]
Hmm, I think it depends on how often these containers may be used together and whether they share common configurations. @eddumelendez , based on your experience with Java, do you think it makes sense to publish them as separate modules or within a single one (the GCloud emulators)? Any opinion regarding the name (`FirestoreContainer` vs. `FirestoreEmulatorContainer`)?

**Comment by dejandjenic** [comment]
basically they don't share configuration at all, for usage point it depends on what are you testing
e.g. in my company we use several levels of testing

if containers are used in unit test then responsibility of certain class is unlikely (should not) involve multiple concerns, repository, service ... should work with firestore or pubsub or bigquery ...
if containers are used for functional (integration) testing via testserver then at server setup it might need to setup entire infrastructure which could need setting multiple containers (here module approach would help)

IMHO i would stick to general testcontainers dotnet design which is one container per service, but this is up to you

on naming i would go with FirestoreContainer , we all know it is an emulator :-)

**Comment by eddumelendez** [comment]
> I think it depends on how often these containers may be used together and whether they share common configurations.

Testcontainers modules based on `google-cloud-cli` image don't share any config. They are group based on the concept of the GCP services.

The relationship between modules and implementation shouldn't be always 1:1. In the future, a `confluentic` module can be shared providing Testcontainers implementations for `schema registry`, `kafka`, `ksqldb` and others. The same for `percona` database which offers images for `mysql`, `postgres`

Feedback about this topic is welcome.

> Any opinion regarding the name (FirestoreContainer vs. FirestoreEmulatorContainer)?

`FirestoreContainer` is ok 👍🏽

**Review by HofmeisterAn — APPROVED** [review]
